Add class HipsSurveyList for parsing list of hips surveys

adl1995 opened the pull request:
I added a class for creating a `astropy.table.Table` from a text file. It reads a list of `OrderedDict` and creates a table from it.

cdeil commented on hips/tiles/surveys.py:
```diff
@@ -0,0 +1,62 @@
+            HiPS list text
+        """
+        surveys = list()
+        print(len(text.split('\n\n')))
```
remove print statement.

cdeil commented on hips/tiles/tests/test_surveys.py:
```diff
@@ -0,0 +1,14 @@
+
+    def test_surveys(self):
+        len_precomp = 335
+        assert len(self.hips_survey_list.surveys) == len_precomp
```
I'd suggest to change the test file to just have a few entries (e.g. 3).
Also, please add an assert or two on some data value in the table, which shows that it was parsed correctly.

cdeil commented on hips/tiles/surveys.py:
```diff
@@ -0,0 +1,62 @@
+        print(len(text.split('\n\n')))
+        for raw_survey in text.split('\n\n'):
+            survey = OrderedDict()
+            for line in raw_survey.split('\n'):
```
Isn't this duplicating the code we already have in https://hips.readthedocs.io/en/latest/api/hips.HipsDescription.html#hips.HipsDescription.parse
Can you call that instead to do the parsing of each survey?

adl1995 commented:
@cdeil - I've made some changes. But, the `parse` method still duplicates the code from `HipsDescription`. Is there any way I can call the `parse` `@classmethod` from within `HipsTile`?

cdeil commented:
My understanding is that the `HipsSurveyList` created here is a list of
https://hips.readthedocs.io/en/latest/api/hips.HipsDescription.html
and the text file format for `HipsSurveyList` is a bunch of text blocks matching the text file format for a single `HipsDescription` concatenated, separated by an empty line.

Is this correct or not?

If yes, you should get the text blocks by splitting on empty lines as you do, and then call `HipsDescription.parse` on each of these text blocks. OK?

adl1995 commented:
@cdeil - I have renamed the class `HipsDescription` and `HipsSurveyList` to `HipsSurveyList` and `HipsSurveyPropertyList`. I've also updated the `parse` method to use `HipsSurveyList.parse`. Please check.

cdeil commented on hips/tiles/tests/test_surveys.py:
```diff
@@ -0,0 +1,22 @@
+        len_precomp = 4
+
+        assert len(self.hips_survey_list.surveys) == len_precomp
+        assert self.hips_survey_list.surveys[0]['surveys']['creator_did'] == 'ivo://CDS/C/MUSE-M42'
```
You could introduce a temp variable here to keep the assert statements nice and short:
```
survey = self.hips_survey_list.surveys[0]['surveys']
```
Also: accessing [0]['surveys'] is very weird, suggest to not put the data in a Table, but just a list of `HipsSurveyProperties` objects.

And please also add a similar assert for some entry other than the first one. E.g. for the last survey you have in that example file.

cdeil commented on hips/tiles/tests/test_surveys.py:
```diff
@@ -0,0 +1,22 @@
+        self.hips_survey_list.fetch()
+
+    def test_surveys(self):
+        len_precomp = 4
```
Please don't use the "precomp" name, it's not very common. Common generic names are "expected" or "desired". In this case though, I think don't need the local variable (just used once, on a line that's already short, just copy the value here).

cdeil commented on hips/tiles/surveys.py:
```diff
@@ -58,6 +58,10 @@ def parse(cls, text: str) -> 'HipsSurveyPropertiesList':
+        return cls(surveys)
+
+    @property
+    def table(self) -> Table:
```
Every public property needs at least a one-line docstring.
And there should be a test covering everything, i.e. in this case for the table.

Also -- I think this way of creating a Table isn't useful. Can you do `table.info()` to show what it contains? Is it a single column? You can't do any selection like that, no?

Probably what you want is to have a table with one row per survey, and the properties as columns.
Can you try `table = Table(rows=rows)` and pass as `rows` a Python list of `OrderedDict` objects?
That should populate the table in a useful way, i.e. with multiple columns and rows as I mentioned above.

If this doesn't work, you could also remove the table property here and then add that in a future PR.

cdeil commented on hips/tiles/surveys.py:
```diff
@@ -14,11 +14,11 @@ class HipsSurveyPropertiesList:
     """
 
-    def __init__(self, surveys: Table) -> None:
+    def __init__(self, surveys: list) -> None:
```
I think  for type annotations to be useful, you'd have to use `from typing import List` and declare the type as `List[HipsSurveyProperties]` here. I've never used that, but maybe you can try?

adl1995 commented:
@cdeil - I have modified the type to `List[HipsSurveyProperties]` in the `__init__` method, but this gives me an error when running `mypy`:

> hips/tiles/surveys.py:61: error: Argument 1 to "HipsSurveyPropertiesList" has incompatible type List[OrderedDict[Any, Any]]; expected List[HipsSurveyProperties]

So, I haven't pushed my changes yet. I don't have much experience with `mypy`, but I tried changing all the type annotations from `HipsSurveyPropertiesList` to `List[HipsSurveyProperties]`. This removes all the errors expect for this one:

> hips/tiles/surveys.py:61: error: Incompatible return value type (got "HipsSurveyPropertiesList", expected List[HipsSurveyProperties])

> hips/tiles/surveys.py:61: error: Argument 1 to "HipsSurveyPropertiesList" has incompatible type List[OrderedDict[Any, Any]]; expected List[HipsSurveyProperties]

cdeil commented on hips/tiles/surveys.py:
```diff
@@ -0,0 +1,67 @@
+        filename : str
+            HiPS list filename
+        """
+        with open(filename, 'r', encoding='utf-8', errors='ignore') as fh:
```
'r' is the default mode for open, no need to specify it.

cdeil commented on hips/tiles/surveys.py:
```diff
@@ -0,0 +1,67 @@
+        url : str
+            HiPS list URL
+        """
+        text = urllib.request.urlopen(url).read()
```
I would suggest to call `decode` instead of calling `str(text), i.e. change to this to be more clear what is happening:
```
response = urllib.request.urlopen(url).read()
text = response.decode('utf-8')
return cls.parse(text)
```

cdeil commented on hips/tiles/surveys.py:
```diff
@@ -0,0 +1,67 @@
+        """
+        surveys = list()
+        for raw_survey in text.split('\n\n'):
+            surveys.append(HipsSurveyProperties.parse(raw_survey).properties)
```
I think the mypy error you mentioned might be due to this line.
You're calling `.properties` here, i.e. storing `OrderedDict` objects in the list.
What do you want to store as data member on `HipsSurveyPropertiesList`.
I would suggest you store `list` of `HipsSurveyProperties`, and not `list` of `OrderedDict`, i.e. don't remove the `.properties` here.

adl1995 commented:
@cdeil - I tried `response.decode('utf-8')`, but this gives me an error:

> UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb2 in position 83958: invalid start byte

Also, I think you meant to remove the `.properties` from `HipsSurveyProperties`, this removes the `mypy` error, but I'm unable to use Astropy `Table` with this.

cdeil commented on hips/tiles/description.py:
```diff
@@ -1,26 +1,38 @@
+    --------
+    ::
+
+        >>> from astropy.utils.data import get_pkg_data_filename
```
You're still using `get_pkg_data_filename` from a docs example here. As discussed previously, this won't work. Please change to a working example where you instead fetch a file from Github or CDS.

cdeil commented on hips/tiles/description.py:
```diff
@@ -2,25 +2,37 @@
+    ::
+
+        >>> import urllib.request
+        >>> response = urllib.request.urlopen('https://raw.githubusercontent.com/hipspy/\
```
This docs example is failing at the moment:
https://gist.github.com/cdeil/422a87d598b1e60e7e5c6bff25d1b716

Please add a `fetch` method to this `HipsSurveyProperties` class and call it from the docs example.

adl1995 commented:
@cdeil - I added the `fetch` classmethod, but also re-based it against `master`, so there are also other files changes showing up.

cdeil commented:
@adl1995 - I think for this PR you also rebased against some old master branch, so now there's a lot of stuff in the diff unrelated to this PR.

Please rebase against latest master and squash all commits into a single one with a good commit message.

cdeil commented:
@adl1995 - I've added a commit in c12acd9ab7797cd7fd7e08a80f0ee802715e95f3 that fixes HipsSurveyPropertiesList.fetch (conversion to text and parsing was broken) and adds a test for that,  adds `from .surveys import *` to `__init__` so that the class shows up in the docs, as well as a bunch of cleanup. I did rename the data members on the `HipsSurveyProperties` and `HipsSurveyPropertiesList` classes to just "data", because I found the previous names confusing because those were often the names of the local variable (i.e. `survey.survey` or `properties.properties` is confusing).

@adl1995 - Will merge soon when travis-ci passes, unless you have any comments.
